docs: Fix broken links in microsite

joaomilho opened the pull request:
## Hey, I just made a Pull Request!

I've removed some links that I found were broken on the documentation.

- Previous reference/createPlugin-router does not exists in https://backstage.io/docs/api/backend
- Previous features/software-catalog/kubernetes-in-backstage does not exists in https://backstage.io/docs/features/software-catalog/software-catalog-api
- Next features/software-catalog/kubernetes-in-backstage does not exists in https://backstage.io/docs/features/software-catalog/external-integrations
- Next reference/createPlugin-router does not exists in https://backstage.io/docs/reference/createPlugin-feature-flags.
- Previous ../CONTRIBUTING does not exists in https://backstage.io/docs/support/support

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
I didn't find the microsite being part of the changeset list. I guess it's because they're no packages or plugins.
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)

<img width="1100" alt="Verify script output with errors" src="https://user-images.githubusercontent.com/308196/108605381-07606f00-73b4-11eb-82a1-f6697a4b9458.png">
<img width="370" alt="Successful verify script output" src="https://user-images.githubusercontent.com/308196/108605384-0af3f600-73b4-11eb-9888-05ff4dc44f10.png">

OrkoHunter commented:
Thanks! I like the idea of verify sidebars check - since it confirms that the rendered "Next" buttons are not broken. I had also found one broken page recently https://github.com/backstage/backstage/pull/4305

(Nit: Renamed `TechDocs:` to `docs:` since `TechDocs:` would mean issues/PRs about the [plugin/core feature](https://backstage.io/docs/features/techdocs/techdocs-overview) and `docs:` would mean anything about the project documentation. 😁 )

Rugvip commented on microsite/sidebars.json:
```diff
@@ -190,7 +188,6 @@
       "architecture-decisions/adrs-adr010",
       "architecture-decisions/adrs-adr011"
     ],
-    "Contribute": ["../CONTRIBUTING"],
```
I think we want this one to work tbh, any idea why it isn't?

joaomilho replied:
Not sure, will investigate.

Rugvip commented on microsite/scripts/verify-sidebars.js:
```diff
@@ -0,0 +1,62 @@
+
+const readMetadata = require('docusaurus/lib/server/readMetadata.js');
+readMetadata.generateMetadataDocs();
+const metadata = require('docusaurus/lib/core/metadata.js');
```
This would be generated as part of `yarn build` in the microsite right? Perhaps we could rely on that instead. Also feels a bit safer to not be running this piece twice 😅

joaomilho replied:
That's true, let me see how to piggyback.

Rugvip reviewed (commented):
Thanks! 🎉

joaomilho commented:
1. The problem with CONTRIBUTING was that docusaurus assumes all files are in the /docs folder. This can be changed (https://docusaurus.io/docs/en/site-config#customdocspath-string) but you can only specify one location. I've created symlinks for CONTRIBUTING and also for ADOPTERS and SECURITY that are linked from CONTRIBUTING. LMK if this solution is problematic.
2. I've removed the bits that built the metadata and moved the workflow step after the build, while adding a proper error if one tries to run the verify:sidebars script without having built.
3. Added some shots in the description above of the output of the script.

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4421-5asn07glir5iwz8buihmtyviugkxhi8l.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60315718fcb22d77055b6f08

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4421-5asn07glir5iwz8buihmtyviugkxhi8l.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60315718fcb22d77055b6f08

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4421-5asn07glir5iwz8buihmtyviugkxhi8l.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60315718fcb22d77055b6f08

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4421-5asn07glir5iwz8buihmtyviugkxhi8l.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60315718fcb22d77055b6f08

adamdmharvey commented on microsite/scripts/verify-sidebars.js:
```diff
@@ -0,0 +1,51 @@
+  const { next, previous } = metadata[id];
+
+  if (next && !ids.includes(next)) {
+    errors.push(`Next ${next} does not exists in ${id}.`);
```
```suggestion
    errors.push(`Next ${next} does not exist in ${id}.`);
```

adamdmharvey commented on microsite/scripts/verify-sidebars.js:
```diff
@@ -0,0 +1,51 @@
+  }
+
+  if (previous && !ids.includes(previous)) {
+    errors.push(`Previous ${previous} does not exists in ${id}.`);
```
```suggestion
    errors.push(`Previous ${previous} does not exist in ${id}.`);
```

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4421-5asn07glir5iwz8buihmtyviugkxhi8l.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60315718fcb22d77055b6f08

joaomilho commented:
Follow up on CONTRIBUTING. We have an either/or situation where we need the file to be at the root (part of gihub mds) or at /docs as part of the microsite. Having in both is not possible as the relative paths will be broken for one of the usages and our original validation script doesn't accept links to the repo.

Other possible solutions:
- Allow full links to the repo and get the symlinks back.
- Move all these to docs (updating relative paths).
- Patching/forking docusaurus to allow multiple root paths?

joaomilho commented:
A bit unsure about the CI failure in verify, it says:

```
Checking formatting...
.github/workflows/microsite-build-check.yml
Code style issues found in the above file(s). Forgot to run Prettier?
error Command failed with exit code 1.
```
It doesn't actually mention any "above files" and it passes locally.

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4421-5asn07glir5iwz8buihmtyviugkxhi8l.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60315718fcb22d77055b6f08

Rugvip commented on .github/workflows/microsite-build-check.yml:
```diff
@@ -43,3 +43,7 @@ jobs:
+
+      - name: verify sidebars
+        run: yarn verify:sidebars
+        working-directory: microsite    
```
Looks like a missing newline might be the problem, set your editor to add trailing newline automatically to avoid! :grin:

joaomilho replied:
wow

Rugvip reviewed (approved):
I think the situation with contributing/contributors as this PR handles it is fine. Seems alright to bounce people by the repo to read the contribution guide tbh

backstage-service commented:
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr4421-5asn07glir5iwz8buihmtyviugkxhi8l.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60315718fcb22d77055b6f08

Rugvip reviewed (commented):
:shipit: 🎉
